File: reference/src/adapters/git-resolver.ts
/**
 * Git-based adapter resolver.
 * Resolves adapter metadata from git repository references (PRD §9.3).
 *
 * Format: `github.com/<org>/<repo>@<ref>`
 * Fetches `metadata.yaml` from the repository's raw content endpoint.
 */

import { parseMetadataYaml } from './scanner.js';
import { validateAdapterMetadata, type AdapterMetadata } from './registry.js';

/** Parsed git adapter reference. */
export interface GitAdapterReference {
  host: string;
  org: string;
  repo: string;
  ref: string;
}

/** Abstraction for fetching raw file content from a git host. */
export interface GitAdapterFetcher {
  fetch(url: string): Promise<string | null>;
}

/** Result of resolving an adapter from a git reference. */
export interface GitResolveResult {
  metadata: AdapterMetadata | null;
  error?: string;
}

const GIT_REF_PATTERN = /^([^/]+)\/([^/]+)\/([^@]+)@(.+)$/;

/**
 * Parse a git adapter reference string.
 * Expected format: `github.com/org/repo@v1.0.0`
 */
export function parseGitAdapterRef(ref: string): GitAdapterReference {
  const match = GIT_REF_PATTERN.exec(ref);
  if (!match) {
    throw new Error(
      `Invalid git adapter reference "${ref}": expected format <host>/<org>/<repo>@<ref>`,
    );
  }
  return {
    host: match[1],
    org: match[2],
    repo: match[3],
    ref: match[4],
  };
}

/**
 * Build the raw content URL for a metadata.yaml file.
 * Currently supports github.com via raw.githubusercontent.com.
 */
export function buildRawUrl(parsed: GitAdapterReference): string {
  if (parsed.host === 'github.com') {
    return `https://raw.githubusercontent.com/${parsed.org}/${parsed.repo}/${parsed.ref}/metadata.yaml`;
  }
  throw new Error(`Unsupported git host: ${parsed.host}`);
}

/**
 * Create a real HTTP fetcher for git-hosted adapter metadata.
 * Uses the global fetch API.
 */
export function createGitAdapterFetcher(): GitAdapterFetcher {
  return {
    async fetch(url: string): Promise<string | null> {
      const response = await globalThis.fetch(url);
      if (!response.ok) return null;
      return response.text();
    },
  };
}

/**
 * Create a stub fetcher for testing.
 * Maps URLs to YAML content strings.
 */
export function createStubGitAdapterFetcher(entries: Map<string, string>): GitAdapterFetcher {
  return {
    async fetch(url: string): Promise<string | null> {
      return entries.get(url) ?? null;
    },
  };
}

/**
 * Resolve adapter metadata from a git reference.
 * Parses the reference, fetches metadata.yaml, validates it.
 */
export async function resolveGitAdapter(
  ref: string,
  fetcher: GitAdapterFetcher,
): Promise<GitResolveResult> {
  let parsed: GitAdapterReference;
  try {
    parsed = parseGitAdapterRef(ref);
  } catch (err) {
    return { metadata: null, error: err instanceof Error ? err.message : String(err) };
  }

  let url: string;
  try {
    url = buildRawUrl(parsed);
  } catch (err) {
    return { metadata: null, error: err instanceof Error ? err.message : String(err) };
  }

  const content = await fetcher.fetch(url);
  if (content === null) {
    return { metadata: null, error: `Failed to fetch metadata from ${url}` };
  }

  let metadata: AdapterMetadata;
  try {
    metadata = parseMetadataYaml(content);
  } catch (err) {
    return {
      metadata: null,
      error: `Invalid YAML: ${err instanceof Error ? err.message : String(err)}`,
    };
  }

  const validation = validateAdapterMetadata(metadata);
  if (!validation.valid) {
    return { metadata: null, error: `Validation failed: ${validation.errors.join('; ')}` };
  }

  return { metadata };
}

File: sdk-python/src/ai_sdlc/adapters/git_resolver.py
"""Git-based adapter resolver.

Resolves adapter metadata from git repository references (PRD Section 9.3).
Format: ``github.com/<org>/<repo>@<ref>``
"""

from __future__ import annotations

import re
from dataclasses import dataclass
from typing import Protocol

from ai_sdlc.adapters.registry import AdapterMetadata, validate_adapter_metadata
from ai_sdlc.adapters.scanner import parse_metadata_yaml

_GIT_REF_PATTERN = re.compile(r"^([^/]+)/([^/]+)/([^@]+)@(.+)$")


@dataclass(frozen=True)
class GitAdapterReference:
    host: str
    org: str
    repo: str
    ref: str


class GitAdapterFetcher(Protocol):
    """Abstraction for fetching raw file content from a git host."""

    async def fetch(self, url: str) -> str | None: ...


@dataclass(frozen=True)
class GitResolveResult:
    metadata: AdapterMetadata | None
    error: str | None = None


def parse_git_adapter_ref(ref: str) -> GitAdapterReference:
    """Parse a git adapter reference string.

    Expected format: ``github.com/org/repo@v1.0.0``
    """
    match = _GIT_REF_PATTERN.match(ref)
    if not match:
        raise ValueError(
            f'Invalid git adapter reference "{ref}": '
            "expected format <host>/<org>/<repo>@<ref>"
        )
    return GitAdapterReference(
        host=match.group(1),
        org=match.group(2),
        repo=match.group(3),
        ref=match.group(4),
    )


def build_raw_url(parsed: GitAdapterReference) -> str:
    """Build the raw content URL for a metadata.yaml file."""
    if parsed.host == "github.com":
        return (
            f"https://raw.githubusercontent.com/"
            f"{parsed.org}/{parsed.repo}/{parsed.ref}/metadata.yaml"
        )
    raise ValueError(f"Unsupported git host: {parsed.host}")


def create_stub_git_adapter_fetcher(entries: dict[str, str]) -> GitAdapterFetcher:
    """Create a stub fetcher for testing. Maps URLs to YAML content strings."""

    class _StubFetcher:
        async def fetch(self, url: str) -> str | None:
            return entries.get(url)

    return _StubFetcher()


async def resolve_git_adapter(
    ref: str,
    fetcher: GitAdapterFetcher,
) -> GitResolveResult:
    """Resolve adapter metadata from a git reference."""
    try:
        parsed = parse_git_adapter_ref(ref)
    except ValueError as err:
        return GitResolveResult(metadata=None, error=str(err))

    try:
        url = build_raw_url(parsed)
    except ValueError as err:
        return GitResolveResult(metadata=None, error=str(err))

    content = await fetcher.fetch(url)
    if content is None:
        return GitResolveResult(
            metadata=None, error=f"Failed to fetch metadata from {url}"
        )

    try:
        metadata = parse_metadata_yaml(content)
    except Exception as err:
        return GitResolveResult(metadata=None, error=f"Invalid YAML: {err}")

    validation = validate_adapter_metadata(metadata)
    if not validation.valid:
        return GitResolveResult(
            metadata=None,
            error=f"Validation failed: {'; '.join(validation.errors)}",
        )

    return GitResolveResult(metadata=metadata)
